File: web/src/core/board/RegionExtractor.ts
import type { BoardTemplate, FieldKey, PixelRect, RowRect } from '../../types/domain';
import { cropCanvas, scaleCanvas } from '../vision/imageUtils';

const FIELD_KEYS: FieldKey[] = ['projectNumber', 'customer', 'daysRemaining'];

/** Minimum crop width fed to OCR - small regions are upscaled for legibility. */
const MIN_OCR_WIDTH = 400;

export interface ExtractedRegion {
  field: FieldKey;
  canvas: HTMLCanvasElement;
  sourceRect: PixelRect;
}

/**
 * Crops only the three configured field regions out of each detected row -
 * never the whole board. Region rectangles are defined as percentages of
 * the row's bounding box in BoardTemplate, so they scale automatically
 * with row height.
 */
export class RegionExtractor {
  private readonly template: BoardTemplate;

  constructor(template: BoardTemplate) {
    this.template = template;
  }

  extractRow(board: HTMLCanvasElement, row: RowRect): ExtractedRegion[] {
    return FIELD_KEYS.map((field) => {
      const frac = this.template.regions[field];
      const sourceRect: PixelRect = {
        x: row.rect.x + frac.xPct * row.rect.width,
        y: row.rect.y + frac.yPct * row.rect.height,
        width: frac.wPct * row.rect.width,
        height: frac.hPct * row.rect.height,
      };
      let canvas = cropCanvas(board, sourceRect);
      if (canvas.width < MIN_OCR_WIDTH) {
        canvas = scaleCanvas(canvas, MIN_OCR_WIDTH / canvas.width);
      }
      return { field, canvas, sourceRect };
    });
  }
}

File: app/src/main/java/com/productionboard/scanner/board/RegionExtractor.kt
package com.productionboard.scanner.board

import android.graphics.Bitmap
import com.productionboard.scanner.domain.BoardTemplate
import com.productionboard.scanner.domain.FieldKey
import com.productionboard.scanner.domain.PixelRect
import com.productionboard.scanner.domain.RowRect

private const val MIN_OCR_WIDTH = 400

data class ExtractedRegion(val field: FieldKey, val bitmap: Bitmap, val sourceRect: PixelRect)

/**
 * Crops only the three configured field regions out of each detected row
 * - never the whole board. Region rectangles are defined as percentages
 * of the row's bounding box in [BoardTemplate], so they scale
 * automatically with row height and with the canonical board resolution
 * configured in Settings/CalibrationScreen.
 */
class RegionExtractor(private val template: BoardTemplate) {

    fun extractRow(board: Bitmap, row: RowRect): List<ExtractedRegion> = FieldKey.entries.map { field ->
        val frac = template.regions.get(field)
        val rect = PixelRect(
            x = row.rect.x + (frac.xPct * row.rect.width).toInt(),
            y = row.rect.y + (frac.yPct * row.rect.height).toInt(),
            width = (frac.wPct * row.rect.width).toInt().coerceAtLeast(1),
            height = (frac.hPct * row.rect.height).toInt().coerceAtLeast(1),
        )
        val clampedX = rect.x.coerceIn(0, (board.width - 1).coerceAtLeast(0))
        val clampedY = rect.y.coerceIn(0, (board.height - 1).coerceAtLeast(0))
        val clampedW = rect.width.coerceAtMost(board.width - clampedX).coerceAtLeast(1)
        val clampedH = rect.height.coerceAtMost(board.height - clampedY).coerceAtLeast(1)

        var cropped = Bitmap.createBitmap(board, clampedX, clampedY, clampedW, clampedH)
        if (cropped.width < MIN_OCR_WIDTH) {
            val scale = MIN_OCR_WIDTH.toFloat() / cropped.width
            cropped = Bitmap.createScaledBitmap(cropped, (cropped.width * scale).toInt(), (cropped.height * scale).toInt(), true)
        }
        ExtractedRegion(field, cropped, rect)
    }
}
